src/setup/node.rs

```diff
@@ -1,193 +1,187 @@
+        self.conf.network_id = Some(network_id);
+        self
+    }
 
-        let (stdout, stderr) = match self.config.log_to_stdout {
+    fn start_node(self, log_to_stdout: bool) -> Node {
+        let (stdout, stderr) = match log_to_stdout {
             true => (Stdio::inherit(), Stdio::inherit()),
             false => (Stdio::null(), Stdio::null()),
         };
-
-        let process = Command::new(&self.meta.start_command)
+        let child = Command::new(&self.meta.start_command)
             .current_dir(&self.meta.path)
             .args(&self.meta.start_args)
             .stdin(Stdio::null())
             .stdout(stdout)
             .stderr(stderr)
             .spawn()
             .expect("node failed to start");
-
-        self.wait_for_start().await;
-        self.process = Some(process);
-
-        Ok(())
-    }
-
-    async fn wait_for_start(&self) {
-        wait_until!(
-            CONNECTION_TIMEOUT,
-            if let Ok(mut stream) = tokio::net::TcpStream::connect(self.addr()).await {
-                stream.shutdown().await.unwrap();
-                true
-            } else {
-                false
-            }
-        );
+        Node {
+            child,
+            addr: self.conf.local_addr,
+        }
     }
+}
 
-    fn generate_config_file(&self) -> Result<()> {
-        let path = self.config.path.join(RIPPLED_CONFIG);
-        let content = RippledConfigFile::generate(&self.config)?;
-        fs::write(path, content)?;
+/// Fields to be written to the node's configuration file.
+#[derive(Debug)]
+pub struct NodeConf {
+    /// The path of the cache directory of the node.
+    pub path: PathBuf,
+    /// The socket address of the node.
+    pub local_addr: SocketAddr,
+    /// The initial peer set of the node.
+    pub initial_peers: HashSet<SocketAddr>,
+    /// The initial max number of peer connections to allow.
+    pub max_peers: usize,
+    /// Token when run as a validator.
+    pub validator_token: Option<String>,
+    /// Network's id to form an isolated testnet.
+    pub network_id: Option<u32>,
+}
 
-        Ok(())
+impl NodeConf {
+    pub fn new(path: PathBuf) -> Self {
+        Self {
+            path,
+            local_addr: SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::LOCALHOST, DEFAULT_PORT)),
+            initial_peers: Default::default(),
+            max_peers: 0,
+            validator_token: None,
+            network_id: None,
+        }
     }
+}
 
-    fn cleanup(&self) -> io::Result<()> {
-        self.cleanup_config_file()?;
-        self.cleanup_cache()
-    }
+pub struct Node {
+    child: Child,
+    addr: SocketAddr,
+}
 
-    fn cleanup_config_file(&self) -> io::Result<()> {
-        let path = self.config.path.join(RIPPLED_CONFIG);
-        match fs::remove_file(path) {
-            // File may not exist, so we suppress the error.
-            Err(e) if e.kind() != std::io::ErrorKind::NotFound => Err(e),
-            _ => Ok(()),
+impl Node {
```

Comment:
I like the builder pattern you used, perhaps having a `builder` function on `Node` that calls the builder's `new` method will make it a little more ergonomic to work with? E.g.:
```
let node = Node::builder(state, path).set_field_smth().set_another_field();
```

Reply:
So `set_field_*` would return builder (in order to be able to chain it)? At what point would `start` be called? I don't know how to achieve above line syntactically, to be honest.